libyara/modules/dotnet/dotnet.c

```diff
@@ -493,6 +498,53 @@ void dotnet_parse_tilde_2(
       row_size = (index_size + (index_sizes.string * 2));
       typeref_row_size = row_size;
       typeref_ptr = table_offset;
+      for (i = 0; i < num_rows; i++)
```

Comment:
I've seen situations in other places where we have a construct like this and it ends up being incorrect. For example, it is possible that we end up setting `number_of_typerefs` to be some non-zero value but never actually populate the array because the string lookups are always returning NULL for some reason. In those cases we have kept a counter to keep track of the number of elements in the array that is independent from the loop counter and use that when setting `number_of_typerefs`.

It has been a while since I wrote this code so I'm not sure if you're just copying my implementation when doing this or not, but it may be worth making a pass through this module to clean this up if it is indeed widespread. If that is the case, then I would suggest doing the cleanup in a follow-up PR.

Reply:
I looked over the four tables this PR is adding: memberref, implmap, typeref, and methoddef. The logic I used for adding a counter was if there is only a name being tracked for the table entry then `counter` was used, if other signatures could be made even with a name not present `i` was used. As mentioned a full sweep of the module could be good in a separate PR with discussion on what warrants the use of a counter with each table.

memberref: used `counter`, only `name` is being tracked for the table
implmap: used `i`, flags alone could be a signature even paired with a null entry name
typeref: used `counter`, only require that name or namespace has an entry to increment counter
methoddef: used `i`, methods also have `rva`, `impl_flags`, and `flags`.